pep-0696.rst

```diff
@@ -0,0 +1,396 @@
+
+   def bar[U = int](): ...
+
+This functionality was included in the initial draft of PEP 695 but was
+removed due to scope creep.
+
+Grammar Changes
+~~~~~~~~~~~~~~~
+
+::
+
+      type_param:
+         | a=NAME b=[type_param_bound] d=[type_param_default]
+         | a=NAME c=[type_param_constraint] d=[type_param_default]
+         | '*' a=NAME d=[type_param_default]
+         | '**' a=NAME d=[type_param_default]
+
+      type_param_default: '=' e=expression
+
+This would mean that ``TypeVarLikes`` with defaults proceeding those
+with non-defaults can be checked at compile time. Although this version
+of the PEP does not define behaviour for ``TypeVarTuple`` and
+``ParamSpec`` defaults, this would mean they can be added easily in the
+future.
+
+Rejected Alternatives
+---------------------
+
+Specification for ``TypeVarTuple`` and ``ParamSpec``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An older version of this PEP included a specification for
+``TypeVarTuple`` and ``ParamSpec`` defaults. However, this has been
+removed as few practical use cases for the two were found. Maybe this
+can be revisited.
+
+Allowing the ``TypeVar`` defaults to be passed to ``type.__new__``\ 's ``**kwargs``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: py
+
+   T = TypeVar("T")
+
+   @dataclass
+   class Box(Generic[T], T=int):
+       value: T | None = None
+
+While this is much easier to read and follows a similar rationale to the
+``TypeVar`` `unary
+syntax <https://github.com/python/typing/issues/813>`__, it would not be
+backwards compatible as ``T`` might already be passed to a
+metaclass/superclass or support classes that don't subclass ``Generic``
+at runtime.
+
+Ideally, if :pep:`637` wasn't rejected, the following would be acceptable:
+
+.. code:: py
+
+   T = TypeVar("T")
+
+   @dataclass
+   class Box(Generic[T = int]):
+       value: T | None = None
+
+Allowing non-defaults to follow defaults
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: py
+
+   YieldT = TypeVar("YieldT", default=Any)
+   SendT = TypeVar("SendT", default=Any)
+   ReturnT = TypeVar("ReturnT")
+
+   class Coroutine(Generic[YieldT, SendT, ReturnT]): ...
+
+   Coroutine[int] == Coroutine[Any, Any, int]
+
+Allowing non-defaults to follow defaults would alleviate the issues with
+returning types like ``Coroutine`` from functions where the most used
+type argument is the last (the return). Allowing non-defaults to follow
+defaults is too confusing and potentially ambiguous, even if only the
+above two forms were valid. Changing the argument order now would also
+break a lot of codebases. This is also solvable in most cases using a
+``TypeAlias``.
+
+.. code:: py
+
+   Coro: TypeAlias = Coroutine[Any, Any, T]
+   Coro[int] == Coroutine[Any, Any, int]
+
+Having ``default`` implicitly be ``bound``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In an earlier version of this PEP, the ``default`` was implicitly set
```

Comment:
More importantly, this would have been a breaking change for existing code. The semantics for unspecified type arguments is well established today, and this would cause that behavior to change.